esper/__init__.py

```diff
@@ -133,19 +136,15 @@ class World:
     frame of your game.
     """
 
-    def __init__(self, timed=False):
-        self._processors = []
-        self._next_entity_id = 0
-        self._components = {}
-        self._entities = {}
-        self._dead_entities = set()
-
-        self._get_component_cache = {}
-        self._get_components_cache = {}
+    def __init__(self) -> None:
+        self._processors: _List[Processor] = []
+        self._next_entity_id: int = 0
+        self._components: _Dict[_Type[_Any], _Set[_Any]] = {}
+        self._entities: _Dict[int, _Dict[_Type[_Any], _Any]] = {}
```

Comment:
Might be that replacing _Any with _C could resolve issue with type: ignore comments

Reply:
``` python
self._entities: _Dict[int, _Dict[_Type[_C], _C] = []
```
```bash
$ mypy esper/
esper/__init__.py:143: error: Type variable "esper._C" is unbound  [valid-type]
esper/__init__.py:143: note: (Hint: Use "Generic[_C]" or "Protocol[_C]" base class to bind "_C" inside a class)
esper/__init__.py:143: note: (Hint: Use "_C" in function signature to bind "_C" inside a function)
Found 1 error in 1 file (checked 1 source file)
```
TypeVar seems to be invalid out of function calls. 
The use of generic (like `class ComponentMap(Generic[_C])`) do not solve this problem because you have to give a type to your Generic (`self._entities: _Dict[int, _Dict[ComponentMap[_Any] = []`).

A solution is to use `assert isinstance(component, component_type)` before return. Then mypy says yes to component is _C. The downside is a tiny performance cost in `__debug__` mode, but I still feel `assert isinstance(...)` more elegant than `# type: ignore`. What do you think of this ?

Reply:
Indeed. 

How about using your typing for `self._entities` and `typing.cast` on returned objects in methods? Like `return cast(_C, self._entities[entity][component_type])`. This should solve need of placing `type: ignore` comments

Reply:
You are right. Thanks you, I find out this function now. I will write a new commit with it tomorrow.

Reply:
Thanks to `typing.cast`, they are no longer `# typing: ignore`. So I file this conversation as "resolved".